src/replace.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use core::fmt::{self, Write};
+use tinyvec::TinyVec;
+
+/// External iterator for replacements for a string's characters.
+#[derive(Clone)]
+pub struct Replacements<I> {
+    iter: I,
+    // At this time, the longest replacement sequence has length 2, so we just
+    // need buffer space for 1 codepoint.
+    buffer: Option<char>,
+}
+
+#[inline]
+pub fn new_svar<I: Iterator<Item = char>>(iter: I) -> Replacements<I> {
+    Replacements { iter, buffer: None }
+}
+
+impl<I: Iterator<Item = char>> Iterator for Replacements<I> {
+    type Item = char;
+
+    #[inline]
+    fn next(&mut self) -> Option<char> {
+        if let Some(c) = self.buffer.take() {
+            return Some(c);
+        }
+
+        match self.iter.next() {
+            Some(ch) => {
+                // At this time, the longest replacement sequence has length 2.
```

Comment:
can we assert this or codegen a constant from the python file?

Reply:
It effectively is asserted by the `TinyVec::<[char; 2]>`, which panics if too many elements are appended. I could codegen the constant if you want, but the code is simpler this way.

Reply:
I could be misreading it, but I thought `TinyVec` allocates if it exceeds the inline array size?

https://docs.rs/tinyvec/1.1.0/tinyvec/enum.TinyVec.html

but yeah, just asserting within the python file should be fine, agreed that codegenerating a constant is probably overkill :)

Reply:
ah yeah, you can use `ArrayVec` which will have the panic on overflow behavior we want here.

Reply:
Oops, looks like I misread the `TinyVec` docs.  Updated to use `ArrayVec`, and I added an assert to the python file.